### ISSUE TITLE
Slow inference relative to Microsoft ONNX Runtime

### ISSUE THREAD

**Issue by joverwey** [snipsco/tract#326]
Running the Squeezenet ONNX model 1000 times using tract-onnx takes ~20 seconds on my machine. Running the same model using the Microsoft ONNX runtime in C++ takes about 9 seconds.

The output results are the same.

At first I thought it was because I have to clone the image on every iteration since the model takes ownership of the input. Why is that btw? Why not allow passing in a reference?

I measured that separately thought and it does not contribute to the overall time.

Here is the Rust code:
```
fn run_image_classification_model() -> TractResult<()> {
    let now = Instant::now();

    let model_path = "d:/temp/model.onnx";

    let shape = tvec!(1, 3, 224, 224);
    let model =
        onnx()
            .model_for_path(model_path)?
            .with_input_fact(0, InferenceFact::dt_shape(f32::datum_type(), shape))?
            .into_optimized()?
            .into_runnable()?;

    eprintln!("Loaded ONNX model");

    let input_tensor_size = 3 * 224 * 224;
    let dummy_data: Vec<f32> =
        (0..input_tensor_size)
            .map(|i| i as f32 / (input_tensor_size + 1) as f32)
            .collect();

    let image: Tensor = Array::from_shape_vec((1, 3, 224, 224), dummy_data)?.into();
    let time_cloning = Instant::now();

    let mut result = model.run(tvec!(image.clone()))?;

    for i in 0..999 {
        result = model.run(tvec!(image.clone()))?;
    }

    println!("Elapsed time: {}", now.elapsed().as_secs());

    for (i, score) in result[0]
        .to_array_view::<f32>()?.iter().take(5).enumerate() {
        println!("Score for class [{}] =  {}\n", i, score);
    }


    Ok(())
}
```

Here is the C++ code:

```
// HeppOnnxCpp.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <onnxruntime_c_api.h>
#include <cstdio>
#include <vector>
#include <cassert>
#include <time.h>
#include <iostream>
#include <iomanip>
using namespace std;

const OrtApi* g_ort = OrtGetApiBase()->GetApi(ORT_API_VERSION);

//*****************************************************************************
// helper function to check for status
void CheckStatus(OrtStatus* status)
{
    if (status != NULL) {
        const char* msg = g_ort->GetErrorMessage(status);
        fprintf(stderr, "%s\n", msg);
        g_ort->ReleaseStatus(status);
        exit(1);
    }
}
int main(int argc, char* argv[]) {
    time_t start, end;
    time(&start);
    //*************************************************************************
    // initialize  enviroment...one enviroment per process
    // enviroment maintains thread pools and other state info
    OrtEnv* env;
    CheckStatus(g_ort->CreateEnv(ORT_LOGGING_LEVEL_WARNING, "test", &env));

    // initialize session options if needed
    OrtSessionOptions* session_options;
    CheckStatus(g_ort->CreateSessionOptions(&session_options));
    g_ort->SetIntraOpNumThreads(session_options, 1);

    // Sets graph optimization level
    g_ort->SetSessionGraphOptimizationLevel(session_options, ORT_ENABLE_BASIC);

    // Optionally add more execution providers via session_options
    // E.g. for CUDA include cuda_provider_factory.h and uncomment the following line:
    // OrtSessionOptionsAppendExecutionProvider_CUDA(sessionOptions, 0);

    //*************************************************************************
    // create session and load model into memory
    // using squeezenet version 1.3
    // URL = https://github.com/onnx/models/tree/master/squeezenet
    OrtSession* session;
#ifdef _WIN32
    const wchar_t* model_path = L"d:/temp/model.onnx";
#else
    const char* model_path = "d:/temp/model.onnx";
#endif

    printf("Using Onnxruntime C API\n");
    CheckStatus(g_ort->CreateSession(env, model_path, session_options, &session));

    //*************************************************************************
    // print model input layer (node names, types, shape etc.)
    size_t num_input_nodes;
    OrtStatus* status;
    OrtAllocator* allocator;
    CheckStatus(g_ort->GetAllocatorWithDefaultOptions(&allocator));

    // print number of model input nodes
    status = g_ort->SessionGetInputCount(session, &num_input_nodes);
    std::vector<const char*> input_node_names(num_input_nodes);
    std::vector<int64_t> input_node_dims;  // simplify... this model has only 1 input node {1, 3, 224, 224}.
                                           // Otherwise need vector<vector<>>

    printf("Number of inputs = %zu\n", num_input_nodes);

    // iterate over all input nodes
    for (size_t i = 0; i < num_input_nodes; i++) {
        // print input node names
        char* input_name;
        status = g_ort->SessionGetInputName(session, i, allocator, &input_name);
        printf("Input %zu : name=%s\n", i, input_name);
        input_node_names[i] = input_name;

        // print input node types
        OrtTypeInfo* typeinfo;
        status = g_ort->SessionGetInputTypeInfo(session, i, &typeinfo);
        const OrtTensorTypeAndShapeInfo* tensor_info;
        CheckStatus(g_ort->CastTypeInfoToTensorInfo(typeinfo, &tensor_info));
        ONNXTensorElementDataType type;
        CheckStatus(g_ort->GetTensorElementType(tensor_info, &type));
        printf("Input %zu : type=%d\n", i, type);

        // print input shapes/dims
        size_t num_dims;
        CheckStatus(g_ort->GetDimensionsCount(tensor_info, &num_dims));
        printf("Input %zu : num_dims=%zu\n", i, num_dims);
        input_node_dims.resize(num_dims);
        g_ort->GetDimensions(tensor_info, (int64_t*)input_node_dims.data(), num_dims);
        for (size_t j = 0; j < num_dims; j++)
            printf("Input %zu : dim %zu=%jd\n", i, j, input_node_dims[j]);

        g_ort->ReleaseTypeInfo(typeinfo);
    }

    // Results should be...
    // Number of inputs = 1
    // Input 0 : name = data_0
    // Input 0 : type = 1
    // Input 0 : num_dims = 4
    // Input 0 : dim 0 = 1
    // Input 0 : dim 1 = 3
    // Input 0 : dim 2 = 224
    // Input 0 : dim 3 = 224

    //*************************************************************************
    // Similar operations to get output node information.
    // Use OrtSessionGetOutputCount(), OrtSessionGetOutputName()
    // OrtSessionGetOutputTypeInfo() as shown above.

    //*************************************************************************
    // Score the model using sample data, and inspect values

    size_t input_tensor_size = 224 * 224 * 3;  // simplify ... using known dim values to calculate size
                                               // use OrtGetTensorShapeElementCount() to get official size!

    std::vector<float> input_tensor_values(input_tensor_size);
    std::vector<const char*> output_node_names = { "softmaxout_1" };

    // initialize input data with values in [0.0, 1.0]
    for (size_t i = 0; i < input_tensor_size; i++)
        input_tensor_values[i] = (float)i / (input_tensor_size + 1);

    // create input tensor object from data values
    OrtMemoryInfo* memory_info;
    CheckStatus(g_ort->CreateCpuMemoryInfo(OrtArenaAllocator, OrtMemTypeDefault, &memory_info));
    OrtValue* input_tensor = NULL;
    CheckStatus(g_ort->CreateTensorWithDataAsOrtValue(memory_info, input_tensor_values.data(), input_tensor_size * sizeof(float), input_node_dims.data(), 4, ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT, &input_tensor));
    int is_tensor;
    CheckStatus(g_ort->IsTensor(input_tensor, &is_tensor));
    assert(is_tensor);
    g_ort->ReleaseMemoryInfo(memory_info);

    // score model & input tensor, get back output tensor
    OrtValue* output_tensor = NULL;
    for (int i = 0; i < 1000; i++) {
        CheckStatus(g_ort->Run(session, NULL, input_node_names.data(), (const OrtValue* const*)&input_tensor, 1, output_node_names.data(), 1, &output_tensor));
    }
    CheckStatus(g_ort->IsTensor(output_tensor, &is_tensor));
    assert(is_tensor);

    // Get pointer to output tensor float values
    float* floatarr;
    CheckStatus(g_ort->GetTensorMutableData(output_tensor, (void**)&floatarr));
    assert(std::abs(floatarr[0] - 0.000045) < 1e-6);

    time(&end);
    // score the model, and print scores for first 5 classes
    for (int i = 0; i < 5; i++)
        printf("Score for class [%d] =  %f\n", i, floatarr[i]);

    // Results should be as below...
    // Score for class[0] = 0.000045
    // Score for class[1] = 0.003846
    // Score for class[2] = 0.000125
    // Score for class[3] = 0.001180
    // Score for class[4] = 0.001317

    g_ort->ReleaseValue(output_tensor);
    g_ort->ReleaseValue(input_tensor);
    g_ort->ReleaseSession(session);
    g_ort->ReleaseSessionOptions(session_options);
    g_ort->ReleaseEnv(env);
    printf("Done!\n");
    // Calculating total time taken by the program. 
    double time_taken = double(end - start);
    cout << "Time taken by program is : " << fixed
        << time_taken << setprecision(5);
    cout << " sec " << endl;
    return 0;
}
```

**Comment by kali** [comment]
Hey ! Thanks for your interest in tract !

About the input, requiring ownerwship of the input is not very satisfactory, but ::run() would need to take it anyway (by making the clone itself). One way to make it a bit nicer would be to allow passing a `impl IntoOwned` maybe. Anyway, as you noticed, copying the input is usually insignificant compared to what's going to happen in the neural network, so I would say this is a low priority item.

About the run performance itself... well maybe MS ONNX Runtime is just faster :) Note that x64 is not my target of choice, I would not be too surprised that they end up being twice faster.

A few things you could look for: 
* tract does not make any effort to run a computation using multiple cores, but is safe to use in multiple threads. So you may get better results by calling run::() on several inputs (or several copies) from different thread (using a parallel iterator may do the trick). I don't know if onnx runtime has some internal multithreading or not.
* I don't know if onnx runtime use GPU acceleration or not, but tract does not, so they may also get an advantage here.

**Comment by kali** [comment]
Just notice that it looks like you actually locked the number of threads to 1, so it's a fair comparison. So they are just faster.

**Comment by joverwey** [comment]
It turns out MS ONNX is using Open MP underneath the hood and it does use multiple cores regardless of setting the number of threads.

I changed the tract example slightly to execute in parallel using rayon:

```
    let results: Vec<_> = (0..1000)
        .into_par_iter()
        .filter_map(|_| model.run(tvec!(image.clone())).ok())
        .collect();

    assert_eq!(results.len(), 1000);
```

Now it executes in < 2 seconds (given 16 cores). Pretty impressive!

**Comment by kali** [comment]
Closing, I don't think there is anything actionable here.

**Comment by kali** [comment]
Mmm... Tensor is Send. TValue is not. Don't convert the tensor `into_tvalue()` out of the closure...

If you can't make it work, please give me the full example, I'll have a look.

**Comment by kali** [comment]
Just realized the full example was up there. Here is how to adapt the old code to the new constraints on tvalue:

```
    let results: Vec<Tensor> = (0..1000)
        .into_par_iter()
        .filter_map(|_| {
            model
                .run(tvec!(image.clone().into_tvalue()))
                .map(|mut res| res.remove(0).into_tensor())
                .ok()
        })
        .collect();
```

**Comment by Christof23** [comment]
Thanks @kali! Just rebenchmarked tract on M1 and averaged over 500 sequential iterations, onnxruntime gets about 60ms/itr, tract is getting 49ms/itr when run in parallel. **Amazing**!

**Comment by kali** [comment]
Wow, that's great news, I did not expect to beat them.

**Comment by kali** [comment]
BTW, if you're curious, there is a WIP for the AMX code there: https://github.com/sonos/tract/pull/956 .

**Comment by Christof23** [comment]
Sounds great, I just have really no idea what's going on under the hood :)

**Comment by Christof23** [comment]
Reconfiguring the code so it will work over a vec of images, with about 45ms/iter.
```
let mut images: Vec<Tensor> = Vec::new();
for i in 0..num_iter {
    images.push(image.clone())
}

let start: Instant = Instant::now();
let results: Vec<Tensor> = images
    .into_par_iter()
    .filter_map(|image| {
        model
            .run(tvec!(image.into_tvalue()))
            .map(|mut res| res.remove(0).into_tensor())
            .ok()
    })
    .collect();
```

**Comment by kali** [comment]
Well, if you want to try the M1 code, just point to the branch in Cargo.toml instead of point versions. It should just pick it up.

**Comment by vishalkumardas** [comment]
@kali How can I implement the same in my code currently, the tract_ONNX takes more than 2 secs in my M1 and the original oNNX takes 345ms to execute. 
Note: The result should always have 7 len because we are predicting 7 fields

fn process_inference() -> Result<Value, Error> {
    let npy_path = "data/preprocessed_data.npy";
    let array_result = read_npy_file(npy_path).map_err(|e| Error::from(e.to_string()))?;

    let expanded_array = array_result.insert_axis(Axis(0));
    let input_data = expanded_array.mapv(|x| x as f32);

    println!("Input data shape: {:?}", input_data.raw_dim());

    let start = Instant::now();

    let model = tract_onnx::onnx()
        .model_for_path("models/test.onnx")?
        .into_optimized()?
        .into_runnable()?;

    let model_load_duration = start.elapsed();
    println!("Load Model time: {:?} seconds", model_load_duration);

    let input_tensor = Tensor::from(input_data);
    println!("Input tensor shape: {:?}", input_tensor.shape());

    let start = Instant::now();
    let result = model.run(tvec!(input_tensor.into()))?;
    let inference_duration = start.elapsed();

    println!("Inference time: {:?} seconds", inference_duration);
    println!("Result length: {}", result.len());

    Ok(json!({ "message": "Inference was successful!" }))
